File: android_app/src/main/java/com/app/remember/DependencyProvider.kt
package com.app.remember

import android.content.Context
import android.content.Intent
import com.app.dependencies.data.dao.RealmDatabase
import com.app.feature.hub.BookmarksAdapter
import com.app.shared.business.AppState
import com.app.shared.business.AppStateReducer
import com.app.shared.data.capture.AndroidDataProcess
import com.app.shared.data.capture.IntentDataCapture
import com.app.shared.data.capture.RawDataCapture
import com.app.shared.data.capture.RawDataProcess
import com.app.shared.data.dao.Database
import com.app.shared.data.repository.BookmarkRepository
import com.app.shared.data.repository.SharedBookmarkRepository
import com.app.shared.feature.mainhub.MainHubViewModel
import com.app.shared.feature.mainhub.SharedMainHubViewModel
import com.app.shared.feature.preview.PreviewViewModel
import com.app.shared.feature.preview.SharedPreviewViewModel
import com.app.shared.navigation.AppNavigation
import com.app.shared.redux.Store
import com.app.shared.utils.CalendarUtils
import com.app.shared.utils.SystemCalendarUtils
import org.koin.dsl.module

class DependencyProvider(private val appContext: Context) {

    val module = module {
        // general
        single { Store(initialState = AppState(), reducer = AppStateReducer) }
        single<Database> { RealmDatabase(context = appContext) }
        single<AppNavigation> { MainAppNavigation() }

        // utils
        single<CalendarUtils> { SystemCalendarUtils() }
        single<RawDataCapture<Intent>> { IntentDataCapture() }
        single<RawDataProcess> { AndroidDataProcess() }

        // repos
        single<BookmarkRepository> {
            SharedBookmarkRepository(
                imageBookmarkDAO = (get() as Database).getImageBookmarkDAO(),
                linkBookmarkDAO = (get() as Database).getLinkBookmarkDAO(),
                textBookmarkDAO = (get() as Database).getTextBookmarkDAO()
            )
        }

        // view models
        factory<PreviewViewModel> {
            SharedPreviewViewModel(
                store = get(),
                bookmarkRepository = get(),
                calendar = get(),
                processor = get())
        }

        factory<MainHubViewModel> {
            SharedMainHubViewModel(
                store = get(),
                calendar = get(),
                bookmarkRepository = get())
        }

        // adapters
        single { BookmarksAdapter() }
    }
}

File: ios_app/ios_dependencies/ios_dependencies/DI/DependencyProvider.swift
//
//  DependencyProvider.swift
//  ios_dependencies
//
//  Created by Gabriel Coman on 14/04/2020.
//  Copyright © 2020 Gabriel Coman. All rights reserved.
//

import Foundation
import RememberShared
import Social
import RealmSwift

@propertyWrapper
public struct Injected<Service> {
    
    public private(set) var wrappedValue: Service
    
    public init() {
        wrappedValue = DependencyProvider.shared.resolve()
    }
}

public typealias DependencyFactory<T> = () -> T

public class DependencyProvider {
    
    public static let shared = DependencyProvider()
    
    private var factories: [ObjectIdentifier: DependencyFactory<Any>] = [:]
    
    public func register<T>(factory: @escaping DependencyFactory<T>) {
        let key = ObjectIdentifier(T.self)
        factories[key] = factory
    }
    
    public func resolve<T>() -> T {
        let key = ObjectIdentifier(T.self)
        
        guard let component = factories[key]?() as? T else {
            fatalError("Dependency '\(T.self)' could not be resolved!")
        }
        
        return component
    }
    
    private init() {
        // the store
        register { ReduxKt.store }
        // the navigations
        // register { AppNavigation() as Navigation }
        
        // the database
        register { RealmDatabase() as Database }
        
        // different utils
        register { SystemDeviceUtils() as DeviceUtils }
        register { SimpleUrlDownloader() as UrlDownloader }
        register { SystemCalendarUtils() as CalendarUtils }
        register { ExtensionContextDataCapture() as RawDataCapture }
        register { SwiftSoupHTMLDataProcess() as HTMLDataProcess }
        register { WebViewProvider() as WebViewProvider }
        register { WebViewContentResolver(provider: self.resolve()) as ContentResolver }
        register { SystemDataProcess(downloader: self.resolve(),
                                  resolver: self.resolve(),
                                  dataProcess: self.resolve()) as RawDataProcess }
        
        // data sources
        register {
            FeedKitItemDataSource() as FeedItemDataSource
        }
        
        // repos
        register {
            SharedBookmarkRepository(
                imageBookmarkDAO: (self.resolve() as Database).getImageBookmarkDAO(),
                textBookmarkDAO: (self.resolve() as Database).getTextBookmarkDAO(),
                linkBookmarkDAO: (self.resolve() as Database).getLinkBookmarkDAO(),
                feedBookmarkDAO: (self.resolve() as Database).getFeedBookmarkDAO()) as BookmarkRepository
        }
        
        register {
            SharedTopicsRepository(topicDAO: (self.resolve() as Database).getTopicDAO()) as TopicsRepository
        }
        
        register {
            SharedFeedItemRepository(dataSource: self.resolve()) as FeedItemRepository
        }
        
        register {
            SharedFeedsRepository(userFeedsDAO: (self.resolve() as Database).getFeedBookmarkDAO(),
                                  allFeedsDAO: DefaultFeedBookmarkDAO()) as FeedsRepository
        }
        
        // view models
        register {
            SharedPreviewViewModel(store: self.resolve(),
                                   bookmarkRepository: self.resolve(),
                                   calendar: self.resolve(),
                                   capture: self.resolve(),
                                   process: self.resolve()) as PreviewViewModel
        }
        
        register {
            SharedBookmarksViewModel(store: self.resolve(),
                                     calendar: self.resolve(),
                                     bookmarkRepository: self.resolve()) as BookmarksViewModel
        }
        register {
            SharedTopicsViewModel(store: self.resolve(),
                                  calendar: self.resolve(),
                                  topicsRepository: self.resolve(),
                                  bookmarkRepository: self.resolve()) as TopicsViewModel
        }
        
        register {
            SharedAddTopicViewModel(store: self.resolve(),
                                    topicsRepository: self.resolve()) as AddTopicViewModel
        }
        
        register {
            SharedEditBookmarkViewModel(store: self.resolve(),
                                        calendar: self.resolve(),
                                        bookmarkRepository: self.resolve(),
                                        topicsRepository: self.resolve()) as EditBookmarkViewModel
        }
        
        register {
            SharedEditTopicViewModel(store: self.resolve(),
                                     topicsRepository: self.resolve()) as EditTopicViewModel
        }
        
        register {
            SharedFeedDetailViewModel(store: self.resolve(),
                                      feedsRepository: self.resolve(),
                                      bookmarkRepository: self.resolve(),
                                      feedItemRepository: self.resolve()) as FeedDetailViewModel
        }
        
        register {
            SharedDisplayViewModel(store: self.resolve(),
                                   bookmarkRepository: self.resolve(),
                                   calendarUtils: self.resolve(),
                                   htmlDataProcess: self.resolve()) as DisplayViewModel
        }
        
        register {
            SharedHubViewModel(store: self.resolve(),
                               calendar: self.resolve(),
                               bookmarkRepository: self.resolve()) as HubViewModel
        }
        
        register {
            SharedFeedsViewModel(store: self.resolve(),
                                 bookmarkRepository: self.resolve(),
                                 calendar: self.resolve()) as FeedsViewModel
        }
    }
}
